docroot/modules/custom/mass_validation/src/Plugin/Validation/Constraint/UnpublishParentConstraintValidator.php

```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace Drupal\mass_validation\Plugin\Validation\Constraint;
+
+use Drupal\Core\StringTranslation\PluralTranslatableMarkup;
+use Drupal\mass_validation\Information\MassChildEntityWarningBuilder;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+
+/**
+ * Checks if an entity has a parent before unpublishing.
+ */
+class UnpublishParentConstraintValidator extends ConstraintValidator {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validate($entity, Constraint $constraint) {
+    if (!isset($entity)) {
+      return;
+    }
+
+    $moderation_state = $entity->get('moderation_state')->value;
+    $unpublished_states = ['unpublished', 'trash'];
+    if (in_array($moderation_state, $unpublished_states)) {
```

Comment:
I prefer early returns, and this line and the line below would be a perfect fit.
Anyhow, the code is easy to read. Good job!

Reply:
Thanks :)